samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+          + "right side of join should be a 'stream'. " + dumpRelPlanForNode(join));
+    }
+
+    validateJoinCondition(join.getCondition());
+  }
+
+  private void validateJoinCondition(RexNode operand) {
+    if (!(operand instanceof RexCall)) {
+      throw new SamzaException("SQL Query is not supported. Join condition operand " + operand +
+          " is of type " + operand.getClass());
+    }
+
+    RexCall condition = (RexCall) operand;
+
+    if (condition.isAlwaysTrue()) {
+      throw new SamzaException("Query results in a cross join, which is not supported. Please optimize the query."
+          + " It is expected that the joins should include JOIN ON operator in the sql query.");
+    }
+
+    if (condition.getKind() != SqlKind.EQUALS && condition.getKind() != SqlKind.AND) {
+      throw new SamzaException("Only equi-joins and AND operator is supported in join condition.");
+    }
+  }
+
+  // Get the stream and table indices corresponding to the fields given in the join condition by parsing through
+  // the condition.
+  private void getStreamAndTableKeyIds(List<RexNode> operands, final LogicalJoin join, boolean isTablePosOnRight,
+      List<Integer> streamIds, List<Integer> tableIds) {
+
+    // All non-leaf operands in the join condition should be expressions.
+    if (operands.get(0) instanceof RexCall) {
+      operands.forEach(operand -> {
+        validateJoinCondition(operand);
+        getStreamAndTableKeyIds(((RexCall) operand).getOperands(), join, isTablePosOnRight, streamIds, tableIds);
+      });
+      return;
+    }
+
+    // We are at the leaf of the join condition. Only binary operators are supported.
+    Validate.isTrue(operands.size() == 2);
+
+    // Only reference operands are supported in row expressions and not constants.
+    // a.key = b.key is supported with a.key and b.key being reference operands.
+    // a.key = "constant" is not yet supported.
+    if (!(operands.get(0) instanceof RexInputRef) || !(operands.get(1) instanceof RexInputRef)) {
+      throw new SamzaException("SQL query is not supported. Join condition " + join.getCondition() + " should have "
+          + "reference operands but the types are " + operands.get(0).getClass() + " and " + operands.get(1).getClass());
+    }
+
+    // Join condition is commutative, meaning, a.key = b.key is equivalent to b.key = a.key.
+    // Calcite assigns the indices to the fields based on the order a and b are specified in
+    // the sql 'from' clause. Let's put the operand with smaller index in leftRef and larger
+    // index in rightRef.
+    RexInputRef leftRef = (RexInputRef) operands.get(0);
+    RexInputRef rightRef = (RexInputRef) operands.get(1);
+    if (leftRef.getIndex() > rightRef.getIndex()) {
+      RexInputRef tmpRef = leftRef;
+      leftRef = rightRef;
+      rightRef = tmpRef;
+    }
+
+    // Get the table index and stream index
```

Comment:
Please rename this table key index and stream key index.

Reply:
Sure